### ISSUE TITLE
New Rule Proposal: jsx-props-prefer-quotes

### ISSUE THREAD

**Issue by karlhorky** [yannickcr/eslint-plugin-react#2868]
Reviewing more code from students today had me thinking: one pattern I see a lot is using a JS expression with only a single string literal inside, as documented in [the React "JSX in Depth" docs](https://reactjs.org/docs/jsx-in-depth.html#string-literals):

```jsx
<Greeting name={'Karl'} />
```

This is in comparison with the more HTML-like (and equivalent):

```jsx
<Greeting name="Karl" />
```

This pattern may or may not be something that individuals and teams wish for, but an important bit for many may be **consistency** across a codebase.

The proposal: a rule called `jsx-props-prefer-quotes` that would do the following for prop expressions containing only a single template string:

### `always`: always prefer props with quotes if possible (my suggestion: make this the default)

Example of **incorrect** code:

```jsx
<Greeting name={'Karl'} />
```

Examples of **correct** code:

```jsx
<Greeting name="Karl" />
```

### `never`: always prefer props with curlies

Example of **incorrect** code:

```jsx
<Greeting name="Karl" />
```

Examples of **correct** code:

```jsx
<Greeting name={'Karl'} />
```

**Comment by karlhorky** [comment]
Ah I guess this has some overlap with [`jsx-no-literals`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-literals.md) with the `ignoreProps` set to `false`, the only feature add being:

- the `always` rule: to instead prefer the literals over expressions

**Comment by ljharb** [comment]
Indeed, this should already be addressed. I've never heard any arguments to force the curly form, so I'm not sure it's worth adding support for that.

**Comment by karlhorky** [comment]
> I've never heard any arguments to force the curly form, so I'm not sure it's worth adding support for that

Oh I think we have a misunderstanding... [the rule currently forces curlies](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-literals.md#rule-options), albeit forcing usage of a variable instead of a string literal:

> Examples of **incorrect** code for this rule, with the above configuration:
> 
> ```jsx
> var Hello = <div class='xx' />;
> var Hello = <div class={'xx'} />;
> ```
>
> Examples of correct **code** for this rule:
> 
> ```jsx
> // use variable for prop values
> var Hello = <div class={xx} />;
> ```

The more I look at this, the more it seems like this `jsx-no-literals` rule has other concerns than what I'm proposing here... The `ignoreProps` could certainly be extended with the thing I'm proposing, but this rule seems to suggest that literals **in any form** are not desirable.

**Comment by ljharb** [comment]
Right - jsx-no-literals is imo about blocking untranslated text.

Are you perhaps looking for https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-brace-presence.md ?

**Comment by karlhorky** [comment]
Ohh, yes, looks like it! 🤦‍♂️

Sorry about that - I tried searching, but I guess for the wrong keywords! Closing...

**Comment by ljharb** [comment]
You may want to look into https://npmjs.com/eslint-config-airbnb, which enables all the helpful rules :-)

**Comment by karlhorky** [comment]
Yeah, I've used that religiously in the past - but now I'm creating one with the rules that I'm teaching at @upleveled, so going through everything now again myself 😅

https://github.com/upleveled/eslint-config-upleveled
